core/arch/arm/pta/sdp_pa.c

```diff
@@ -0,0 +1,105 @@
+/*
+ * Copyright (C) 2017, ARM Limited, All Rights Reserved
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include <kernel/pseudo_ta.h>
+#include <mm/core_memprot.h>
+#include <mm/tee_mmu.h>
+#include <sdp_pa.h>
+#include <string.h>
+
+#define PTA_NAME "sdp_pa.pta"
+
+/*
+ * Specificed TA needs to create pagetable item for data
+ * put in SDP memory. Create a PTA for specficed TA to retrieve
+ * SDP memory physical address.
+ */
+static TEE_Result sdp_pa_cmd_virt_to_phys(uint32_t types,
+			     TEE_Param params[TEE_NUM_PARAMS])
+{
+	char *va = params[0].memref.buffer;
+	size_t len = params[0].memref.size;
+	paddr_t pa;
+	struct tee_ta_session *s;
+	struct user_ta_ctx *utc;
+	TEE_Result res;
+
+	if (types != TEE_PARAM_TYPES(TEE_PARAM_TYPE_MEMREF_INPUT,
+				     TEE_PARAM_TYPE_VALUE_OUTPUT,
+				     TEE_PARAM_TYPE_NONE,
+				     TEE_PARAM_TYPE_NONE)) {
+		EMSG("bad parameters types: %x", (unsigned int)types);
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
+
+	s = tee_ta_get_calling_session();
+	if (!s)
+		return TEE_ERROR_ACCESS_DENIED;
+
+	utc = to_user_ta_ctx(s->ctx);
+	res = tee_mmu_check_access_rights(utc, TEE_MEMORY_ACCESS_READ |
+				TEE_MEMORY_ACCESS_WRITE |
+				TEE_MEMORY_ACCESS_ANY_OWNER,
+				(uaddr_t)va, len);
+	if (res != TEE_SUCCESS)
+		return TEE_ERROR_ACCESS_DENIED;
+
+	if (!core_vbuf_is(CORE_MEM_SDP_MEM, va, len)) {
+		EMSG("bad memref secure attribute");
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
+
+	pa = virt_to_phys(va);
+
+	params[1].value.a = (uint32_t)pa;
+	params[1].value.b = (uint32_t)(pa >> 32);
+
+	return TEE_SUCCESS;
+}
+
+/*
+ * Trusted Application Entry Points
+ */
+static TEE_Result open_session(uint32_t nParamTypes __unused,
+		TEE_Param pParams[TEE_NUM_PARAMS] __unused,
+		void **ppSessionContext __unused)
+{
+	struct tee_ta_session *s = tee_ta_get_calling_session();
+
+	/*
+	 * Check that we're called from a TA
+	 * with TA_FLAG_SECURE_DATA_PATH flag
+	 */
+	if (s && (s->ctx->flags & TA_FLAG_SECURE_DATA_PATH)) {
+		DMSG("open entry point for pseudo ta \"%s\"", PTA_NAME);
+		return TEE_SUCCESS;
+	}
+
+	EMSG("unauthorised access to pseudo ta \"%s\"", PTA_NAME);
+
+	return TEE_ERROR_ACCESS_DENIED;
+}
+
+static TEE_Result invoke_command(void *pSessionContext __unused,
+		uint32_t nCommandID, uint32_t nParamTypes,
+		TEE_Param pParams[TEE_NUM_PARAMS])
+{
+	FMSG("command entry point for pseudo ta \"%s\"", PTA_NAME);
+
+	switch (nCommandID) {
+	case PTA_CMD_SDP_VIRT_TO_PHYS:
+		return sdp_pa_cmd_virt_to_phys(nParamTypes, pParams);
+	default:
+		break;
+	}
+
+	return TEE_ERROR_BAD_PARAMETERS;
```

Comment:
s/TEE_ERROR_BAD_PARAMETERS/TEE_ERROR_NOT_SUPPORTED/
or
s/TEE_ERROR_BAD_PARAMETERS/TEE_ERROR_NOT_IMPLEMENTED/

Reply:
I am prefer to use TEE_ERROR_NOT_IMPLEMENTED am will update it too.